Wrap captured variable with WeakPtr in GPUProcess::initializeGPUProcess

jace0x21 opened the pull request:
#### c27c56bff56cb60c7e717975cfe3b8cf1eec31c0
<pre>
Wrap captured variable with WeakPtr in GPUProcess::initializeGPUProcess
<a href="https://bugs.webkit.org/show_bug.cgi?id=254316">https://bugs.webkit.org/show_bug.cgi?id=254316</a>
rdar://107117855

Reviewed by Ryosuke Niwa and Youenn Fablet.

We are storing a lambda with a captured variable. We should use a
smart pointer here so we can check if this is alive before accessing it.

* Source/WebKit/GPUProcess/GPUProcess.cpp:
(WebKit::GPUProcess::initializeGPUProcess):

Canonical link: <a href="https://commits.webkit.org/262314@main">https://commits.webkit.org/262314@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cdbf986449104d9868b75f38ca1ad46a758e9bda

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/942 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/968 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/1006 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/1379 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/836 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/1020 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/1054 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/1053 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/951 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/911 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/886 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/1293 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/936 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/883 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/882 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/859 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/910 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/1919 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/900 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/853 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/847 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/892 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/300 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/914 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebKit/GPUProcess/GPUProcess.cpp:
```diff
@@ -236,9 +236,11 @@ void GPUProcess::initializeGPUProcess(GPUProcessCreationParameters&& parameters)
-        parentProcessConnection()->sendWithAsyncReply(Messages::GPUProcessProxy::StatusBarWasTapped(), [] { }, 0);
-        completionHandler();
+    CoreAudioSharedUnit::unit().setStatusBarWasTappedCallback([process = WeakPtr<GPUProcess> { *this }](auto completionHandler) {
+        if (process) {
```
We should store this process in RefPtr like so:
```cpp
if (RefPtr process = weakProcess.get()) {
    process->parentProcessConnection()...
}
```

youennf commented on Source/WebKit/GPUProcess/GPUProcess.cpp:
```diff
@@ -236,9 +236,11 @@ void GPUProcess::initializeGPUProcess(GPUProcessCreationParameters&& parameters)
-    CoreAudioSharedUnit::unit().setStatusBarWasTappedCallback([this](auto completionHandler) {
-        parentProcessConnection()->sendWithAsyncReply(Messages::GPUProcessProxy::StatusBarWasTapped(), [] { }, 0);
-        completionHandler();
+    CoreAudioSharedUnit::unit().setStatusBarWasTappedCallback([weakProcess = WeakPtr<GPUProcess> { *this }](auto completionHandler) {
```
```suggestion
    CoreAudioSharedUnit::unit().setStatusBarWasTappedCallback([weakProcess = WeakPtr { *this }] (auto completionHandler) {
```

youennf commented on Source/WebKit/GPUProcess/GPUProcess.cpp:
```diff
@@ -236,9 +236,11 @@ void GPUProcess::initializeGPUProcess(GPUProcessCreationParameters&& parameters)
+    CoreAudioSharedUnit::unit().setStatusBarWasTappedCallback([weakProcess = WeakPtr<GPUProcess> { *this }](auto completionHandler) {
+        if (RefPtr process = weakProcess.get()) {
+            process->parentProcessConnection()->sendWithAsyncReply(Messages::GPUProcessProxy::StatusBarWasTapped(), [] { }, 0);
+            completionHandler();
```
We need to call completionHandler outside the if statement.

jace0x21 committed:
Wrap captured variable with WeakPtr in GPUProcess::initializeGPUProcess
https://bugs.webkit.org/show_bug.cgi?id=254316
rdar://107117855

Reviewed by Ryosuke Niwa and Youenn Fablet.

We are storing a lambda with a captured variable. We should use a
smart pointer here so we can check if this is alive before accessing it.

* Source/WebKit/GPUProcess/GPUProcess.cpp:
(WebKit::GPUProcess::initializeGPUProcess):

Canonical link: https://commits.webkit.org/262314@main
